python_modules/dagster/dagster_tests/core_tests/resource_tests/pythonic_resources/test_direct_invocation.py

```diff
@@ -427,11 +428,9 @@ def my_other_asset(context):
     my_asset(ctx)
     assert ctx.get_output_metadata("result") == {"foo": "bar"}
 
-    with pytest.raises(
-        DagsterInvariantViolationError,
-        match="attempted to log metadata for output 'result' more than once",
-    ):
-        my_other_asset(ctx)
+    # context in unbound when used in another invocation. This allows the metadata to be
+    # added in my_other_asset
+    my_other_asset(ctx)
 
 
 def test_async_assets_with_shared_context():
```

Comment:
the context is getting unbound then rebound with the info for `async_asset_two`, so the assert in `async_asset_one` is failing. This means anyone who is testing their assets/ops with the same context and running the ops synchronously will be getting bad info in some of the ops. need to figure out a solution to this

Reply:
@alangenfeld  I remember talking about this with you, did we decide that we wouldn't support this case anymore, or that there was a way to support it?

Reply:
my recollection is that we concluded its probably acceptable to not support concurrent use of the same context

based on your description it sounds like we can probably raise a more explicit error if we knew if the context was still in use [2]

Reply:
the current implementation logs a warning message if a bound context is reused. I'm not sure how to detect the difference between reusing a context for assets run in series (so we can warn) vs reusing a context for assets run via `async` (so we can fail, or do something else)

Reply:
the thought i had was that if we modeled the direct invocation context so that we knew when it was each of the states of

`unbound` -> `bound / pre-execution` -> `post-execution`

then you could distinguish attempting to re-bind on a post-execution context versus attempting to re-bind on a pre-execution context 

the simplest way would be a setting a `complete` bool or something
a fancier way might be related to how `user_events` / `output_metadata` are set-up since those are dynamically mutated during invocation but then should theoretically be read only after that

Reply:
i think with the `complete` bool we run into the same issue as trying to automatically `unbind` the context after execution is complete. There isn't a place in the code path that has access to the context when the execution is complete (that i'm aware of). Maybe we could add something to the `observe_output` function where if all of the expected outputs of the op have been "seen" we set a `completed` bool?

Reply:
>  There isn't a place in the code path that has access to the context when the execution is complete (that i'm aware of)

maybe in or near `_type_check_output_wrapper` ? maybe with some slight modifications?